### PR TITLE
Elasticache support

### LINKED ISSUE #207: Support for elasticache
Elasticache provisioning support would be nice.

http://docs.aws.amazon.com/AWSRubySDK/latest/AWS/ElastiCache/Client/V20140930.html

### PR DESCRIPTION
**Pull request by dblessing**
Fixes #207 - Support for defining an Elasticache cluster (Redis/Memcache)

- [x] Subnet group
* ~~Security group~~
- [x] Cache cluster
- [x] Replication group (Required for redis clustering)

Edit: Security groups aren't necessary. Only if creating a cache cluster outside of a VPC

Like RDS, Elasticache has a one to many relationship between subnet/security groups and the actual cache cluster. Thus, they must be separate resources. Unfortunately it is not possible to re-use regular VPC subnet/security groups.

Elasticache is kind of a beast because there's no SDK resource class, only a client. This implementation works for subnet groups just fine. Hopefully we can carry that in to the security groups and cache cluster, too. 

Input is greatly appreciated. I decided to go w/ `cache` in the resource names as opposed to `elasticache` to keep it from being too wordy. 

I will add some tests before I'm done.

Subnet group resource example (vpc is implicit based on subnets specified):
```
aws_cache_subnet_group 'test-ec' do
  description 'My awesome group'
  subnets [ 'public-test' ]
end
```

Cache cluster resource example:
```
aws_cache_cluster 'my-cluster-mem' do
  number_nodes 2
  node_type 'cache.t2.micro'
  engine 'memcached'
  engine_version '1.4.14'
  security_groups ['my_security_group']
  subnet_group_name 'test-ec'
end
```

### CODE REVIEW

File: lib/chef/resource/aws_elasticache_subnet_group.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  attribute :aws_tags # hack until bug is fixed https://github.com/chef/chef-provisioning-aws/issues/204
+
+  def aws_object
+    begin
+      driver.elasticache
+        .describe_cache_subnet_groups(cache_subnet_group_name: name)
+        .data[:cache_subnet_groups].first
+    rescue AWS::ElastiCache::Errors::CacheSubnetGroupNotFoundFault
+      nil
+    end
```
Comment (dblessing) [inline lib/chef/resource/aws_elasticache_subnet_group.rb]: The above seems totally insane, but from my understanding of the SDK this is the best I can come up with.

File: lib/chef/resource/aws_cache_cluster.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  # Tags are actually allowed on the cluster
+  attribute :aws_tags
+
+  def aws_object
+    begin
+      cluster = driver.elasticache
+        .describe_cache_clusters(cache_cluster_id: cluster_name)
+        .data[:cache_clusters].first
+      Chef::Log.warn(cluster.inspect)
+      cluster
```
Comment (dblessing) [inline lib/chef/resource/aws_cache_cluster.rb]: Remove debug bits

File: lib/chef/provider/aws_elasticache_subnet_group.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    @desired_options ||= begin
+      options = {}
+      options[:cache_subnet_group_name] = new_resource.name
+      options[:cache_subnet_group_description] = new_resource.description
+      options[:subnet_ids] = new_resource.subnets
+      AWSResource.lookup_options(options, resource: new_resource)
+    end
+  end
+
+  def update_required(elasticache_subnet_group)
```
Comment (dblessing) [inline lib/chef/provider/aws_elasticache_subnet_group.rb]: change to `update_required?`

File: lib/chef/provider/aws_cache_cluster.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  private
+
+  # def update_cache_subnet_group
+  #   new_resource.driver.elasticache.modify_cache_subnet_group(desired_options)
+  # end
+
+  def desired_options
+    @desired_options ||= begin
+      options = {}
+      options[:cache_cluster_id] = new_resource.cluster_name
```
Comment (metadave) [inline lib/chef/provider/aws_cache_cluster.rb]: is `cache_cluster_id` from `cluster_name` intentional?

Comment (dblessing) [inline lib/chef/provider/aws_cache_cluster.rb]: Yes, in this case the 'id' is a friendly name instead of the normal aws `sg-1234abcd`. I could change the attribute to be `cluster_id` but `cluster_name` felt more indicative of what it can accept.

Comment (metadave) [inline lib/chef/provider/aws_cache_cluster.rb]: sounds good, thanks

File: lib/chef/provider/aws_cache_replication_group.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  def create_aws_object
+    converge_by "create new Elasticache Replication Group #{new_resource.name} in #{region}" do
+      new_resource.driver.elasticache.create_replication_group(desired_options)
+    end
+  end
+
+  def update_aws_object(cache_replication_group)
+    Chef::Log.warn('Updating Elasticache Replication Groups is currently unsupported')
+  end
```
Comment (dblessing) [inline lib/chef/provider/aws_cache_replication_group.rb]: The `describe_replication_group` method is so limiting in what it returns. It's almost impossible for us to tell if attributes have changed. It's also complicated because a replication group actually launches `n` cache clusters.

**Comment by dblessing** [comment]
This is probably not ready for merge, but it's ready for feedback. All the basic elements are here for spinning up cache clusters. I wrote tests for what I could, but the nature of how cache clusters are spun up limits testing within the current framework. The issue comes down to the `after` block trying to destroy the cache cluster that was just created. Since the cluster is in `creating` state you can't destroy until it's fully available.

What do you think about the implementation? If you like it I can squash commits and it'll be ready for merge.

File: lib/chef/provider/aws_cache_replication_group.rb (hunk 5 above)
Comment (metadave) [inline lib/chef/provider/aws_cache_replication_group.rb]: can you add this as a comment to the method so context isn't lost once the PR is merged?

**Comment by metadave** [comment]
looks good, I would love to see examples and/or docs.

File: lib/chef/provider/aws_cache_replication_group.rb (hunk 5 above)
Comment (dblessing) [inline lib/chef/provider/aws_cache_replication_group.rb]: I think I'll change my statement after more thought and research - It's *possible* to get info about each cache cluster and do recursive updates but it's going to take a lot more logic and API calls. For the sake of getting basic support out there I think we should move forward with this PR and then come back around and add support for this and additional attributes.

Comment (tyler-ball) [inline lib/chef/provider/aws_cache_replication_group.rb]: Yeah, I'm fine with that - we can have a later PR to add update functionality in.

File: lib/chef/resource/aws_cache_cluster.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+require 'chef/provisioning/aws_driver/aws_resource'
+
+class Chef::Resource::AwsCacheCluster < Chef::Provisioning::AWSDriver::AWSResource
+  aws_sdk_type AWS::ElastiCache
+
+  attribute :cluster_name, kind_of: String, name_attribute: true
+  attribute :az_mode, kind_of: String
+  attribute :preferred_availability_zone, kind_of: String
+  attribute :preferred_availability_zones,
```
Comment (tyler-ball) [inline lib/chef/resource/aws_cache_cluster.rb]: Why do we need both `:prefered_availability_zone` and `:prefered_availability_zones` if the plural version can also take in a String?  Seems like its just more work to have to do a diff later to see which the user provided and whether one has priority over the other or should we thrown an error if both are provided, etc.

File: lib/chef/resource/aws_cache_cluster.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
+require 'chef/provisioning/aws_driver/aws_resource'
+
+class Chef::Resource::AwsCacheCluster < Chef::Provisioning::AWSDriver::AWSResource
+  aws_sdk_type AWS::ElastiCache
+
+  attribute :cluster_name, kind_of: String, name_attribute: true
+  attribute :az_mode, kind_of: String
```
Comment (tyler-ball) [inline lib/chef/resource/aws_cache_cluster.rb]: I know its more verbose, but I think the readability is worth it for `:availability_zone_mode`

File: lib/chef/resource/aws_cache_cluster.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+require 'chef/provisioning/aws_driver/aws_resource'
+
+class Chef::Resource::AwsCacheCluster < Chef::Provisioning::AWSDriver::AWSResource
+  aws_sdk_type AWS::ElastiCache
```
Comment (tyler-ball) [inline lib/chef/resource/aws_cache_cluster.rb]: I think its worth a comment explaining that there isn't actually a resource class

File: lib/chef/resource/aws_cache_cluster.rb (hunk 6 above)
Comment (dblessing) [inline lib/chef/resource/aws_cache_cluster.rb]: I tried to stay as close to the API implementation as possible. Both are present here http://docs.aws.amazon.com/AWSRubySDK/latest/AWS/ElastiCache/Client/V20140930.html#create_cache_cluster-instance_method 

I can try to combine them, but I need to test what valid values are for each. The `preferred_availability_zones` documentation makes the array values look weird.

**Comment by dblessing** [comment]
Thanks for the feedback @tyler-ball

File: lib/chef/resource/aws_cache_subnet_group.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+require 'chef/provisioning/aws_driver/aws_resource'
+
+class Chef::Resource::AwsCacheSubnetGroup < Chef::Provisioning::AWSDriver::AWSResource
+  aws_sdk_type AWS::ElastiCache, id: :group_name
```
Comment (tyler-ball) [inline lib/chef/resource/aws_cache_subnet_group.rb]: Did you find a reason you needed to set the `id` field here?  I think its only used for the option lookup, which it doesn't seem like we need for these resources.

File: spec/integration/aws_cache_subnet_group_spec.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
+require 'spec_helper'
+
+describe Chef::Resource::AwsCacheSubnetGroup, skip_after: true do
```
Comment (tyler-ball) [inline spec/integration/aws_cache_subnet_group_spec.rb]: Why are you `skip_after`ing here?

File: spec/integration/aws_cache_subnet_group_spec.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  extend AWSSupport
+
+  when_the_chef_12_server "exists", organization: 'foo', server_scope: :context do
+    with_aws "with a VPC with an internet gateway and subnet" do
+      before :context do
+        driver.ec2.vpcs.with_tag('Name', 'test_vpc').each do |vpc|
```
Comment (tyler-ball) [inline spec/integration/aws_cache_subnet_group_spec.rb]: This should be happening automatically for you - did you not see that behavior?

File: lib/chef/provider/aws_cache_subnet_group.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+
+class Chef::Provider::AwsCacheSubnetGroup < Chef::Provisioning::AWSDriver::AWSProvider
+
+  protected
+
+  def create_aws_object
+    converge_by "create new Elasticache Subnet Group #{new_resource.name} in #{region}" do
+      new_resource.driver.elasticache.create_cache_subnet_group(desired_options)
```
Comment (tyler-ball) [inline lib/chef/provider/aws_cache_subnet_group.rb]: I would make a getter for `new_resource.driver.elasticache` so you don't have to type so much :smile:

File: lib/chef/resource/aws_cache_subnet_group.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+  attribute :group_name, kind_of: String, name_attribute: true
+  attribute :description, kind_of: String, required: true
+  attribute :subnets,
+            kind_of: [ String, Array ],
```
Comment (tyler-ball) [inline lib/chef/resource/aws_cache_subnet_group.rb]: You don't need the `LookupOptions` you have in the provider unless you also add `AwsSubnet, AWS::EC2::Subnet` like in the [aws_network_interface](https://github.com/chef/chef-provisioning-aws/blob/master/lib/chef/resource/aws_network_interface.rb#L14)

**Comment by tyler-ball** [comment]
I think this is looking really great @dblessing - awesome job.  I had a few notes/questions but the majority of it looks really good.

I think the only things missing now are a `ref` file which contains examples of the resources fully populated and integration tests.

:cake: for you

File: lib/chef/resource/aws_cache_subnet_group.rb (hunk 13 above)
Comment (dblessing) [inline lib/chef/resource/aws_cache_subnet_group.rb]: Great catch. I do want to resolve `AwsSubnet, AWS::EC2::Subnet`.

File: lib/chef/resource/aws_cache_cluster.rb (hunk 6 above)
Comment (tyler-ball) [inline lib/chef/resource/aws_cache_cluster.rb]: <sigh> Because AWS supports both attributes, it makes sense for us to allow both.  The plural version is also only supported for Memcached.  Lets just keep both here and pass them through to AWS, then AWS can bubble up any errors.

**Comment by tyler-ball** [comment]
Hey @dblessing - I see this hasn't been updated in a while.  Do you need help with it, or have you just been busy?  Thanks for the contributions so far - its a lot of really great work, so let us know if you need help getting it across the finish line!

**Comment by dblessing** [comment]
@tyler-ball I've just been busy. I will try to get it finished up this week. Thanks for your feedback.

File: lib/chef/resource/aws_cache_subnet_group.rb (hunk 9 above)
Comment (dblessing) [inline lib/chef/resource/aws_cache_subnet_group.rb]: Yes, it didn't work without this. It gave me an error about the 'name' attribute.

File: spec/integration/aws_cache_subnet_group_spec.rb (hunk 11 above)
Comment (dblessing) [inline spec/integration/aws_cache_subnet_group_spec.rb]: Just copied from another test file. I'll remove and see if it works.

**Comment by dblessing** [comment]
@tyler-ball I think it's ready for a final look

**Comment by tyler-ball** [comment]
Hey @dblessing - thanks for all the work on this.  We're going to assign this to @metadave to get it over the finish line (more specs if possible, updates to `ref_full.rb` and a `ref_elasticache.rb` example doc) before getting it merged.

**Comment by metadave** [comment]
I'm trying to put together a `ref_elasticache.rb` file:

```
require 'chef/provisioning/aws_driver'
with_driver 'aws'

aws_security_group 'ec-test-security-group' do
  inbound_rules '0.0.0.0/0' => [ 22, 80 ]
end

aws_vpc "ec-test-vpc" do
  cidr_block '10.0.0.0/24'
  internet_gateway true
end

aws_subnet "ec-test-subnet" do
  vpc 'ec-test-vpc'
  availability_zone 'us-east-1a'
  cidr_block "10.0.0.0/24"
end

aws_cache_subnet_group 'ec-test-subnet-group' do
  description 'Test elasticache subnet group'
  subnets [ 'ec-test-subnet' ]
end

aws_cache_cluster 'my-cluster-mem' do
  preferred_availability_zone 'us-east-1a'
  number_nodes 2
  node_type 'cache.t2.micro'
  engine 'memcached'
  engine_version '1.4.14'
  security_groups ['ec-test-security-group']
  subnet_group_name 'ec-test-subnet-group'
end
```

but it bombs out with:

```
Error executing action `create` on resource 'aws_cache_cluster[my-cluster-mem]'
...
Subnet group [ec-test-subnet-group] belongs to a different VPC [vpc-d6f4f4b3] than [vpc-64519c01]
```
looks like `vpc-64519c01` is a default with a bunch of Chef instances attached to it. Am I missing an attribute somewhere?

**Comment by dblessing** [comment]
@metadave lgtm. I just ran the following test (same I used when developing this):

```
require 'chef/provisioning/aws_driver'
with_driver 'aws::us-east-1'

aws_vpc 'test' do
  cidr_block "10.0.0.0/24"
end

aws_subnet "public-test" do
  vpc 'test'
  availability_zone 'us-east-1a'
  cidr_block "10.0.0.0/24"
end

aws_cache_subnet_group 'test-ec' do
  description 'My awesome group'
  subnets [ 'public-test' ]
end

aws_security_group 'test-sg' do
  vpc 'test'
end

aws_cache_cluster 'my-cluster-mem' do
  az_mode 'single-az'
  number_nodes 2
  node_type 'cache.t2.micro'
  engine 'memcached'
  engine_version '1.4.14'
  security_groups ['test-sg']
  subnet_group_name 'test-ec'
end
```

**Comment by metadave** [comment]
thanks, that works for me too.

**Comment by metadave** [comment]
This is awesome, I'm going to merge and submit a few cleanup PR's with examples, specs and docs (docs.chef.io). Thanks @dblessing

**Comment by dblessing** [comment]
Thanks @metadave.

**Comment by metadave** [comment]
np. I'm making sure the specs work correctly before merging.

On Tue, Jun 16, 2015 at 7:29 PM, Drew Blessing <notifications@github.com>
wrote:

> Thanks @metadave <https://github.com/metadave>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/chef/chef-provisioning-aws/pull/212#issuecomment-112601803>
> .
>



-- 

*Dave Parfitt* — Engineering

716.218.3213– dparfitt@chef.io - my: Linkedin
<http://www.linkedin.com/in/metadave/> Twitter
<http://www.twitter.com/metadave>

*CHEF*

TM

*chef.io* <http://www.chef.io/> Blog <http://www.opscode.com/blog/> Facebook
<https://www.facebook.com/getchefdotcom> Twitter <https://twitter.com/chef>
Youtube <https://www.youtube.com/getchef>
